### PR TITLE
DTensor impl of aten.native_dropout_backward.default

### PR DESCRIPTION
**Pull request by crutcher**

### CODE REVIEW

File: spmd/tensor/ops/dropout.py [hunk 1]
```diff
@@ -28,3 +28,31 @@ def _dist_dropout(
                 mask, device_mesh=self.device_mesh, placements=self.placements
             ),
         )
+
+@register_impl("aten.native_dropout_backward.default")
```
Comment (wanchaol) [inline spmd/tensor/ops/dropout.py:32]: I know this is WIP, but can we use sharding propagation rule instead of register_impl? `register_impl` is only supposed to be used by custom written implementation. Example usage of sharding propagation rule https://github.com/pytorch/tau/blob/main/spmd/tensor/ops/math_ops.py#L34

**Comment by crutcher** [comment]
@wanchaol can you trigger the workflows on this? It's ready for review

File: test/spmd/tensor/test_pointwise_ops.py [hunk 2]
```diff
@@ -10,50 +13,182 @@
...
+                device_mesh=device_mesh,
+                placements=placements,
+            )
+            for k, v in val.items()
+        }
+
+    return val
+
+
+def collect_dtensor(val: Any) -> Any:
```
Comment (anj-s) [inline test/spmd/tensor/test_pointwise_ops.py:70]: nit: I would rename the functions to "convert_to_dtensor" and "convert_from_dtensor" for easy readability.

Comment (crutcher) [reply test/spmd/tensor/test_pointwise_ops.py]: Done; `deepcopy_convert_{to,from}_dtensor()`

File: spmd/tensor/ops/dropout.py (hunk 1 above)
Comment (crutcher) [reply spmd/tensor/ops/dropout.py]: done, using the built-in delegation mechanisms now.

**Review by anj-s — APPROVED** [review]

File: test/spmd/tensor/test_pointwise_ops.py [hunk 3]
```diff
@@ -10,50 +13,182 @@
...
+) -> Any:
+    """
+    Recursively coerce (over Sequence and Dict types) Tensors into DTensors.
+
+    :param device_mesh: the DeviceMesh to use.
+    :param placements: the Placement list to use.
+    :return: the transformed structure.
+    """
+    if isinstance(val, Tensor):
+        return DTensor(
```
Comment (wanchaol) [inline test/spmd/tensor/test_pointwise_ops.py:40]: could you use `distribute_tensor` API instead? https://github.com/pytorch/tau/blob/main/spmd/tensor/__init__.py#L14 This is a more legit API to use (i.e. shard an existing torch.Tensor to different ranks), instead of creating a DTensor from each rank's local tensor.

File: spmd/requirements_dev.txt [hunk 4]
```diff
@@ -1,9 +1,12 @@
...
 black
-expecttest
 flake8
 numpy
 pyre-check
 pytest
 pytest-cov
 pytest-xdist
+
+expecttest
```
[the remark this answers is not in the document]
Comment (crutcher) [reply spmd/requirements_dev.txt:10]: done; though ... opinions.

File: test/spmd/tensor/test_pointwise_ops.py [hunk 5]
```diff
@@ -10,50 +13,182 @@
...
 from torch.testing._internal.common_distributed import skip_if_lt_x_gpu
 
 
+def deepcopy_convert_to_dtensor(
+    val: Any,
+    device_mesh: DeviceMesh,
+    placements: Sequence[Placement],
+) -> Any:
+    """
+    Recursively coerce (over Sequence and Dict types) Tensors into DTensors.
```
[the remark this answers is not in the document]
Comment (crutcher) [reply test/spmd/tensor/test_pointwise_ops.py:33]: done

File: test/spmd/tensor/test_pointwise_ops.py (hunk 3 above)
Comment (crutcher) [reply test/spmd/tensor/test_pointwise_ops.py:40]: done.

Comment (anj-s) [reply test/spmd/tensor/test_pointwise_ops.py]: @wanchaol Should we add a comment about what is the suggested way to create a DTensor and what the benefits of using distribute_tensor are? I feel I've run into this issue of using DTensor directly as well.

File: test/spmd/tensor/test_pointwise_ops.py [hunk 6]
```diff
@@ -8,52 +11,184 @@
...
+        reset_seed: Optional[Callable] = None,
+        args: Sequence[Any] = tuple(),
+        kwargs: Optional[Dict[str, Any]] = None,
+    ):
+        if not kwargs:
+            kwargs = {}
+
+        torch.manual_seed(self.rank)
+
+        dargs = deepcopy_convert_to_dtensor(
```
Comment (aazzolini) [inline test/spmd/tensor/test_pointwise_ops.py:105]: would pytree help here? I think you can do 

  import torch._pytree as pytree

  flattened_args = pytree.tree_flatten(args)
  dtensor_args  = [convert_to_dtensor(x) for x in flattened_args]
  d_args = pytree.tree_unflatten(flattened_args)

or :

   d_args = pytree.tree_map(args, convert_to_dtensro)

File: test/spmd/tensor/test_pointwise_ops.py (hunk 3 above)
Comment (wanchaol) [reply test/spmd/tensor/test_pointwise_ops.py]: yeah sure, let me update the comment of Tensor constructor to reflect this.

File: spmd/requirements_dev.txt (hunk 4 above)
[the remark this answers is not in the document]
Comment (wanchaol) [reply spmd/requirements_dev.txt]: could you share your opinions :) i think you might want to separate the requirements to different sections that have different type of deps, i think it's fine to categorize them, we just need to add comments to distinguish.

**Review by aazzolini — COMMENTED** [review]
So what is the status right now, is dropout the only test that is failing right now ?

File: test/spmd/tensor/test_pointwise_ops.py (hunk 6 above)
Comment (crutcher) [reply test/spmd/tensor/test_pointwise_ops.py:93]: done

**Review by wanchaol — CHANGES_REQUESTED** [review]
Left some questions about the change, and specifically there seem to be other changes to CI and device mesh, which it's not related to this PR and need further discussions imo, we should remove unrelated changes before merging.

File: .github/workflows/code-quality.yml [hunk 7]
```diff
@@ -15,30 +15,34 @@ jobs:
...
         pip install -r docs/requirements.txt
         pip install types-docutils types-setuptools tqdm
-        if [ -f requirements.txt ]; then pip install -r requirements.txt --find-links https://download.pytorch.org/whl/nightly/cpu/torch_nightly.html; fi
+        if [ -f requirements.txt ]; then
+            pip install -r requirements.txt --find-links https://download.pytorch.org/whl/nightly/cpu/torch_nightly.html;
+        fi
         pip install --pre torchvision -f https://download.pytorch.org/whl/nightly/cpu/torch_nightly.html
         pip install git+https://github.com/pbelevich/transformers.git@compatible_with_pt_master
         pip install "black<23" pylint==v3.0.0a5 mypy==v0.902 flake8 pyre-check==0.9.15
+
```
[the remark this answers is not in the document]
Comment (crutcher) [reply .github/workflows/code-quality.yml:34]: the if condition doesn't do anything; and mypy was breaking

File: spmd/requirements_dev.txt [hunk 8]
```diff
@@ -2,8 +2,10 @@
 black
 expecttest
 flake8
+hypothesis
 numpy
 pyre-check
 pytest
 pytest-cov
 pytest-xdist
+pyyaml
```
[the remark this answers is not in the document]
Comment (crutcher) [reply spmd/requirements_dev.txt:11]: because it's used in the tests already and tests fail without it.

File: spmd/tensor/device_mesh.py [hunk 9]
```diff
@@ -40,6 +40,16 @@ def set_global_device_mesh(mesh: Optional["DeviceMesh"]) -> None:
     _global_device_mesh = mesh
 
 
+T = TypeVar("T")
+_L = Union[T, Sequence[T]]
+DeepList = _L[_L[_L[_L[_L[int]]]]]
```
[the remark this answers is not in the document]
Comment (crutcher) [reply spmd/tensor/device_mesh.py:45]: This was failing mypy typechecking on construction.

The type is meant to be "n-dimensional int Tensors, and things that torch.as_tensor(t) will convert cleanly to n-dimensional int tensors".

The later cannot be expressed cleanly in mypy types; so the choice was to remove the annotation (or use Any), or to do this.

This is the standard way to simulate recursive structure types in mypy. It is not as general as true recursive type support, but this will handle up to 5-dimensional expressions.

File: test/spmd/tensor/test_pointwise_ops.py [hunk 10]
```diff
@@ -8,94 +12,270 @@
...
+    def _compare_pairwise_ops(
+        self,
+        *,
+        device_mesh: DeviceMesh,
+        placements: Sequence[Placement],
+        op: Callable,
+        pre_op_reset: Optional[Callable] = None,
+        args: Sequence[Any] = tuple(),
+        kwargs: Optional[Dict[str, Any]] = None,
+        verify_args: bool = True,
```
[the remark this answers is not in the document]
Comment (crutcher) [reply test/spmd/tensor/test_pointwise_ops.py:85]: it's being used everywhere, as this is the default behavior; but it's expensive and the option will permit us to turn it off. This is calling out that that branch isn't required.

File: test/spmd/tensor/test_pointwise_ops.py [hunk 11]
```diff
@@ -8,94 +12,270 @@
...
+
 
 class DistElementwiseOpsTest(DistTensorTestBase):
+    def _compare_pairwise_ops(
+        self,
+        *,
+        device_mesh: DeviceMesh,
+        placements: Sequence[Placement],
+        op: Callable,
+        pre_op_reset: Optional[Callable] = None,
```
[the remark this answers is not in the document]
Comment (crutcher) [reply test/spmd/tensor/test_pointwise_ops.py:82]: this is a generalized refactoring of code that was already there. calling a reset callback is strictly more flexible than hardcoding some specialized behavior, and we should always select the composition answer over the specialized case answer.

File: test/spmd/tensor/test_pointwise_ops.py [hunk 12]
```diff
@@ -8,94 +12,270 @@
...
+
+    :param val: the structure to coerce.
+    :return: the coerced structure.
+    """
+
+    def f(x):
+        if isinstance(x, DTensor):
+            return x.redistribute(
+                device_mesh=x.device_mesh,
+                placements=[Replicate()],
```
[the remark this answers is not in the document]
Comment (crutcher) [reply test/spmd/tensor/test_pointwise_ops.py:68]: Can you explain how you would consolidate results locally? Because this is what I was given an an answer on the chat

File: spmd/tensor/device_mesh.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (wanchaol) [reply spmd/tensor/device_mesh.py:45]: `torch.tensor(mesh, dtype=torch.int)` is literally do the same thing as `torch.as_tensor` by "convert" cleanly, and that does not require to have this pseudo recursive type definition for mypy, what I am wondering is that why we need to switch to `torch.as_tensor`, I don't see a strong reason on this, that's why I think we need more discussion on it. Also this seems not relating to this PR, if you want to change this, feel free to open another PR and we can discuss further

File: test/spmd/tensor/test_pointwise_ops.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (wanchaol) [reply test/spmd/tensor/test_pointwise_ops.py:68]: `placements=[Replicate()] * x.device_mesh.ndim` should be sufficient

File: test/spmd/tensor/test_pointwise_ops.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (wanchaol) [reply test/spmd/tensor/test_pointwise_ops.py:82]: `pre_op_reset` seems to be not a generalized refactoring, it looks to me just a random seed "reset" name because there's no other reset can be done effectively, if you want to generalize it, maybe rename it to sth like `pre_op_fn`.

File: test/spmd/tensor/test_pointwise_ops.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (wanchaol) [reply test/spmd/tensor/test_pointwise_ops.py:85]: My point is that there's no callsite to turn this off, and I think we shouldn't over engineering, we could introduce this when it have the case to turn it off if you found it's expensive. In fact, I can't see why need to turn it off as it always started with torch.Tensor as args/kwargs and we need to convert them to DTensor in order to execute it.

File: spmd/tensor/device_mesh.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (crutcher) [reply spmd/tensor/device_mesh.py:45]: the type that DeviceMesh took was _already_ broken. It worked in one very narrow range, that of `[[int]]`; it didn't handle `[int]` or `[[[int]]]`; and it was throwing mypy errors when I used it in testing.

I rewrote the type annotation to fix this, rather than putting `# type: ignore` at the call site.

The constructor cleanup was not needed, I was just already there, and `.as_tensor()` is cleaner than what was being done (and we already weren't doing a clean copy, so I fixed that too).

This change is more correct, and more flexible.

Are you arguing against doing this, or that these cleanups should be in seperate PRs?

[the remark this answers is not in the document]
Comment (wanchaol) [reply spmd/tensor/device_mesh.py:45]: I think it's good to fix the type annotation, i'm not against that, but there're several options to avoid this DeepList annotations. Also I am curious why it breaks the test because I manually create a 3d int list and mypy didn't complain about it at all. I think It should at least in this separate PR and we can discuss further there, i.e. mypy released their experimental recursive types support https://mypy-lang.blogspot.com/2022/09/mypy-0981-released.html Maybe we want to use this for a cleaner type annotation

File: test/spmd/tensor/test_pointwise_ops.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (crutcher) [reply test/spmd/tensor/test_pointwise_ops.py]: done.

File: spmd/tensor/device_mesh.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (crutcher) [reply spmd/tensor/device_mesh.py]: Lifted out into seperate PR: https://github.com/pytorch/tau/pull/515

File: test/spmd/tensor/test_pointwise_ops.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (crutcher) [reply test/spmd/tensor/test_pointwise_ops.py]: done.

File: test/spmd/tensor/test_pointwise_ops.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (crutcher) [reply test/spmd/tensor/test_pointwise_ops.py]: done.

**Review by wanchaol — COMMENTED** [review]
Looking good, only have a few comments.

File: test/spmd/tensor/test_pointwise_ops.py [hunk 13]
```diff
@@ -8,94 +12,269 @@
...
+        dkwargs = deepcopy_convert_to_dtensor(
+            kwargs,
+            device_mesh=device_mesh,
+            placements=placements,
+        )
+
+        # These could be skipped.
+        self.assertEqual(
+            args,
+            deepcopy_convert_from_dtensor(args),
```
[the remark this answers is not in the document]
Comment (crutcher) [reply test/spmd/tensor/test_pointwise_ops.py:106]: yes; fixed.

File: test/spmd/tensor/test_pointwise_ops.py [hunk 14]
```diff
@@ -8,94 +12,269 @@
...
+                device_mesh=x.device_mesh,
+                placements=[Replicate()] * x.device_mesh.ndim,
+            ).to_local()
+        return x
+
+    return pytree.tree_map(f, [val])[0]
+
 
 class DistElementwiseOpsTest(DistTensorTestBase):
+    def _compare_pairwise_ops(
```
[the remark this answers is not in the document]
Comment (crutcher) [reply test/spmd/tensor/test_pointwise_ops.py:76]: both get used in existing tests; one generates input; one just runs what it gets.

I'll be building more tests after this on these.

File: test/spmd/tensor/test_pointwise_ops.py [hunk 15]
```diff
@@ -8,94 +12,269 @@
...
+            placements=placements,
+        )
+        dkwargs = deepcopy_convert_to_dtensor(
+            kwargs,
+            device_mesh=device_mesh,
+            placements=placements,
+        )
+
+        # These could be skipped.
+        self.assertEqual(
```
[the remark this answers is not in the document]
Comment (crutcher) [reply test/spmd/tensor/test_pointwise_ops.py:104]: These were the things I made conditional before, with the `if validate_args:` section.

I can remove them.

File: test/spmd/tensor/test_pointwise_ops.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (wanchaol) [reply test/spmd/tensor/test_pointwise_ops.py:76]: I don't get why we need two tbh, we can just move the the input generation `torch.randn` to `_compare_pairwise_ops` function, no?

btw I think the `_compare_pairwise_ops` maybe renaming to `_compare_pointwise_ops`?

**Review by wanchaol — APPROVED** [review]
lgtm
